solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1588,12 +1589,36 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
       if (updateLastState) {
         cd.getCloudDescriptor().setLastPublished(state);
       }
-      overseerJobQueue.offer(Utils.toJSON(m));
+      DocCollection coll = zkStateReader.getCollection(collection);
+      if(forcePublish || sendToOverseer(coll, coreNodeName)) {
+        overseerJobQueue.offer(Utils.toJSON(m));
+      } else {
+        if(log.isDebugEnabled()) {
+          log.debug("bypassed overseer for message : {}", Utils.toJSONString(m));
+        }
+        PerReplicaStates perReplicaStates = coll.getPerReplicaStates(zkClient);
+        PerReplicaStates.WriteOps ops = PerReplicaStates.WriteOps.flipState(coreNodeName, state, perReplicaStates);
+        PerReplicaStates.persist(ops, perReplicaStates.path, zkClient);
+      }
     } finally {
       MDCLoggingContext.clear();
     }
   }
 
+  private boolean sendToOverseer(DocCollection coll, String coreNodeName) {
+    if (coll == null || !coll.isPerReplicaState()) return true;
```

Comment:
nit: But what is Solr clouds code style regarding conditionals and curlies being optional? Granted in this instance all of these are single return statements. So not a huge issue. But multi-statement conditionals without curlies has bitten so many people before that we enforce them in our Java code style at FullStory.

Reply:
the coding style allows no curlies for simple `if` checks